retro/utils/geom.py

```diff
@@ -390,9 +401,9 @@ def generate_digitizer(bin_edges):
         )
         def digitize(val):
             if val < start:
-                return -1
-            if val >= stop:
-                return num_bin_edges
+                return underflow_idx
+            if val > stop:
```

Comment:
It is my intention to keep the behavior same as numpy, i.e., values at the last bin edge are excluded from the last bin. This is the logic I've used elsewhere, so if there's compelling reason to change, need to make sure logic is modified everywhere; otherwise, change `>` to `>=`. As you've added the option to the user to "clip" values already, seems like if a user needs upper-edge-inclusiveness (all I can think of are for azimuth and coszen angles), using the `clip=True` should do the trick.

Reply:
(If you do need the upper-edge to be inclusive, then update the `test_generate_digitizer` unit test to make sure this passes)